flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -887,20 +887,28 @@ private ApplicationReport startAppMaster(
 			}
 		}
 
-		// setup security tokens
 		Path remotePathKeytab = null;
+		String localizedKeytabPath = null;
 		String keytab = configuration.getString(SecurityOptions.KERBEROS_LOGIN_KEYTAB);
 		if (keytab != null) {
-			LOG.info("Adding keytab {} to the AM container local resource bucket", keytab);
-			remotePathKeytab = setupSingleLocalResource(
-					Utils.KEYTAB_FILE_NAME,
+			boolean	requireLocalizedKeytab = flinkConfiguration.getBoolean(YarnConfigOptions.SHIP_LOCAL_KEYTAB);
+			localizedKeytabPath = flinkConfiguration.getString(YarnConfigOptions.LOCALIZED_KEYTAB_PATH);
+			if (requireLocalizedKeytab) {
+				// Localize the keytab to YARN containers via local resource.
+				LOG.info("Adding keytab {} to the AM container local resource bucket", keytab);
+				remotePathKeytab = setupSingleLocalResource(
+					localizedKeytabPath,
 					fs,
 					appId,
 					new Path(keytab),
 					localResources,
 					homeDir,
 					"",
 					fileReplication);
+			} else {
```

Comment:
Aren't we doing duplicate work here? If we don't have a local keytab and we don't ship it, won't the configuration be already in place and the cluster entrypoint can use it?

Reply:
+1